crates/sui-core/src/authority_aggregator.rs

```diff
@@ -961,46 +971,32 @@ where
-                                            &mut state, signed, name, weight,
-                                        )
-                                    {
-                                        return ReduceOutput::Success(cert);
-                                    }
-                                }
-                                VerifiedTransactionInfoResponse::Executed(cert, effects) => {
-                                    if let Some(result) = self
-                                        .handle_response_with_certified_transaction(
-                                            &mut state, cert, effects, name, weight,
-                                        )
-                                    {
-                                        return ReduceOutput::Success(result);
-                                    }
-                                }
-                            },
-                            Err(err) => {
-                                self.handle_response_with_err(
-                                    &mut state, name, weight, tx_digest, err,
-                                );
+                        match self.handle_process_transaction_response(
+                            tx_digest, &mut state, response, name, weight,
+                        ) {
+                            Ok(Some(result)) => {
+                                self.record_process_transaction_metrics(tx_digest, &state);
+                                ReduceOutput::Success(result)
                             }
-                        };
+                            Ok(None) => ReduceOutput::Continue(state),
+                            Err(err) => {
+                                debug!(?tx_digest, name=?name.concise(), weight, "Error processing transaction from validator: {:?}", err);
+                                self.metrics
+                                    .process_tx_errors
+                                    .with_label_values(&[&name.concise().to_string(), err.as_ref()])
+                                    .inc();
+                                state.bad_stake += weight;
+                                state.errors.push((err, vec![name], weight));
 
-                        if state.bad_stake > validity {
-                            self.metrics
-                                .num_signatures
-                                .observe(state.signatures.len() as f64);
-                            self.metrics.num_good_stake.observe(state.good_stake as f64);
-                            self.metrics.num_bad_stake.observe(state.bad_stake as f64);
-                            return ReduceOutput::Failed(state);
+                                if state.bad_stake > validity_threshold {
+                                    ReduceOutput::Failed(state)
+                                } else {
+                                    ReduceOutput::Continue(state)
```

Comment:
nit: we could put these lines into a function of StakeAggregator and return the output result

Reply:
I don't know what that means. StakeAggregator is only aggregating one thing, while here bad_stake could come from anything, including things that are not even added into the StakeAggregator.